packages/i18n-calypso/types/index.d.ts

```diff
@@ -0,0 +1,87 @@
+// Type definitions for i18n-calypso
+// Project: i18n-calypso
+
+/**
+ * External dependencies
+ */
+import { ComponentClass, ComponentType, ReactElement } from 'react';
+import moment from 'moment';
+
+declare namespace i18nCalypso {
+	export type Substitution = string | number;
+
+	export type Substitutions =
+		| Substitution
+		| Substitution[]
+		| { [placeholder: string]: Substitution };
+
+	export interface ComponentInterpolations {
+		[placeholder: string]: ReactElement;
+	}
+
+	export interface TranslateOptions {
+		/**
+		 * Arguments you would pass into sprintf to be run against the text for string substitution.
+		 */
+		args?: Substitutions;
+
+		/**
+		 * Components for interpolation in the provided string.
+		 */
+		components?: ComponentInterpolations;
+
+		/**
+		 * Comment that will be shown to the translator for anything that may need to be explained about the translation.
+		 */
+		comment?: string;
+
+		/**
+		 * Provides the ability for the translator to provide a different translation for the same text in two locations (dependent on context). Usually context should only be used after a string has been discovered to require different translations. If you want to provide help on how to translate (which is highly appreciated!), please use a comment.
+		 */
+		context?: string;
+	}
+
+	export function translate( original: string ): string;
+	export function translate(
+		original: string,
+		plural: string,
+		options: TranslateOptions & { count: number }
+	): string;
```

Comment:
Thanks for bringing that up! You pointed me to a few things:

`translate` is hookable 😱
It would be good to add a type for translation hooks so that they conform

Indeed, when components are involved, we use the return type of `interpolate-components`… I'll see about adding a declaration for that. It seems to return a `ReactFragment`:

https://github.com/Automattic/interpolate-components/blob/cdc890587b826e6ff0aa5e4868bf37c29d093931/src/index.es6#L94

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3620be6751c532f1c3ce08d03554ee273fa14f70/types/react-addons-create-fragment/index.d.ts#L11